### PR TITLE
Playwright video tests added - Start/End time - Transcript time

### LINKED ISSUE #1127: Playwright tests for video start/end times.
As a followup to #1124, we should add some playwright tests for video pages. Some things we could check:

- start/end time on videos are passed correctly to youtube *Look at the iframe's `src` property*
- start/end times reflected properly in expandable transcript (on video pages, not on embeds)
- iframe src has `controls` as expected.

### PR DESCRIPTION
**Pull request by fakhar-ud-din**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #1127

#### What's this PR do?
This PR adds playwright tests in hugo-themes for video start and end times as well as start time of the transcripts.

#### How should this be manually tested?
- Checkout to this branch
- Run the following command in ocw-hugo-themes,
   - yarn test:e2e
- All tests should be successful

### CODE REVIEW

**Review by ChristopherChudzicki — CHANGES_REQUESTED** [review]
**Overall Comment:** Nice. And very cool that we can use Playwright to interact with the Iframe. I wasn't expecting that to work. I don't think we should be manipulating the iframe a whole lot in our tests, but since the Iframe controls code we do own (e.g., transcript), it seems appropriate.

**Requested Change:** Now that we understand the play button issue better, let's do the appropriate waiting without an explicit timeout.

**Suggestion:** I'd also like to suggest that you organize some of the video interaction code into a ["Page object model"](https://playwright.dev/docs/pom). Maybe something like this:

```ts
class VideoSection {
  private readonly page: Page

  constructor(page: Page) {
    this.page = page
  }

  /**
   * Returns a playwright locator for the video section of the page, containing
   * the video player itself as well as transcript / expandable tabs.
   * 
   * @fakhar goal here is to scope other the selectors to this part of the DOM.
   */
  videoSection(): Locator {
    return this.page.locator(".video-page")
  }

  /**
   * @returns Returns a video section Iframe.
   */
  iframe(): Locator {
    return this.videoSection().locator(
      'role=region[name="Video Player"] >> iframe'
    )
  }

  frameLocator(): FrameLocator {
    return this.iframe().frameLocator(":scope")
  }

  /**
   * @fakhar This is one way we can wait for the video player to be ready.
   * Use as `await videoSectionInstance().expectPlayerReady()`.
   *
   * NOTE!!! Requires updating Playright; 1.27 does not have expect(...).toPass()
   * as far as I can tell.
   */
  async expectPlayerReady(): Promise<void> {
    await expect(async () => {
      const readyState = await this.frameLocator()
        .locator("*")
        .first()
        .evaluate(() => window.document.readyState)
      expect(readyState).toEqual("complete")
    }).toPass()
  }

  playButton(): Locator {
    return this.frameLocator().getByRole("button", { name: "Play" })
  }

  transcript = {
    lines:      () => this.videoSection().locator(".transcript-line"),
    activeLine: () => this.transcript.lines().locator(".is-active")
  }
}
```

A few reasons I think the "page object model" approach is worthwhile:
1. automate common tasks (waiting for iframe to be ready could be common)
2. Gives a natural place to add explanatory comments (an explanation about `expectPlayerReady` is definitely worthwhile.)
3. Isolates CSS selectors. In general, I think we should avoid CSS-based Playwright selectors (preferring `byRole`). But sometimes they are most convenient. Like asserting about `.transcript-line` But CSS selectors are more fragile. Putting the CSS Selectors in a centralized place makes easier to fix issues down the road if classnames change or something.

**Review by ChristopherChudzicki — CHANGES_REQUESTED** [review]
**Overall Comment:** Nice. And very cool that we can use Playwright to interact with the Iframe. I wasn't expecting that to work. I don't think we should be manipulating the iframe a whole lot in our tests, but since the Iframe controls code we do own (e.g., transcript), it seems appropriate.

**Requested Change:** Now that we understand the play button issue better, let's do the appropriate waiting without an explicit timeout.

**Suggestion:** I'd also like to suggest that you organize some of the video interaction code into a ["Page object model"](https://playwright.dev/docs/pom). Maybe something like this:

```ts
class VideoSection {
  private readonly page: Page

  constructor(page: Page) {
    this.page = page
  }

  /**
   * Returns a playwright locator for the video section of the page, containing
   * the video player itself as well as transcript / expandable tabs.
   * 
   * @fakhar goal here is to scope other the selectors to this part of the DOM.
   */
  videoSection(): Locator {
    return this.page.locator(".video-page")
  }

  /**
   * @returns Returns a video section Iframe.
   */
  iframe(): Locator {
    return this.videoSection().locator(
      'role=region[name="Video Player"] >> iframe'
    )
  }

  frameLocator(): FrameLocator {
    return this.iframe().frameLocator(":scope")
  }

  /**
   * @fakhar This is one way we can wait for the video player to be ready.
   * Use as `await videoSectionInstance().expectPlayerReady()`.
   *
   * NOTE!!! Requires updating Playright; 1.27 does not have expect(...).toPass()
   * as far as I can tell.
   */
  async expectPlayerReady(): Promise<void> {
    await expect(async () => {
      const readyState = await this.frameLocator()
        .locator("*")
        .first()
        .evaluate(() => window.document.readyState)
      expect(readyState).toEqual("complete")
    }).toPass()
  }

  playButton(): Locator {
    return this.frameLocator().getByRole("button", { name: "Play" })
  }

  transcript = {
    lines:      () => this.videoSection().locator(".transcript-line"),
    activeLine: () => this.transcript.lines().locator(".is-active")
  }
}
```

A few reasons I think the "page object model" approach is worthwhile:
1. automate common tasks (waiting for iframe to be ready could be common)
2. Gives a natural place to add explanatory comments (an explanation about `expectPlayerReady` is definitely worthwhile.)
3. Isolates CSS selectors. In general, I think we should avoid CSS-based Playwright selectors (preferring `byRole`). But sometimes they are most convenient. Like asserting about `.transcript-line` But CSS selectors are more fragile. Putting the CSS Selectors in a centralized place makes easier to fix issues down the road if classnames change or something.

File: tests-e2e/ocw-ci-test-course/shortcodes.spec.ts [hunk 1]
```diff
@@ -37,3 +37,75 @@ test("Related resources link to correct page", async ({ page }) => {
...
+  await course.goto("resources/ocw_test_course_mit8_01f16_l01v01_360p")
+  const src = await page.locator("iframe.vjs-tech").getAttribute("src")
+
+  const urlParams = new URLSearchParams(src || "")
+  expect(urlParams.get("start")).toEqual(expectedStartTime)
+
+  await page
+    .frameLocator('role=region[name="Video Player"] >> iframe')
+    .locator(".ytp-cued-thumbnail-overlay-image")
+    .getAttribute("style")
```
[the remark this answers is not in the document]
Comment (fakhar-ud-din) [reply tests-e2e/ocw-ci-test-course/shortcodes.spec.ts]: This was added to ensure that the video thumbnail loads, indicating player has finished loading.

File: tests-e2e/ocw-ci-test-course/shortcodes.spec.ts [hunk 2]
```diff
@@ -37,3 +37,68 @@ test("Related resources link to correct page", async ({ page }) => {
...
+  expect(urlParams.get("end")).toEqual(expectedEndTime)
+})
+
+test("Transcripts start time matches video start time", async ({ page }) => {
+  const course = new CoursePage(page, "course")
+  const videoSection = new VideoSection(page)
+  const expectedStartTime = "13"
+
+  await course.goto("resources/ocw_test_course_mit8_01f16_l01v01_360p")
+  const src = await page.locator("iframe.vjs-tech").getAttribute("src")
```
Comment (ChristopherChudzicki) [inline tests-e2e/ocw-ci-test-course/shortcodes.spec.ts:83]:
```suggestion
  const src = await videoSection.iframe().getAttribute("src")
```

Nicer to not use the `.vjs-tech` class. *There's still a class in the `VideoSection` class used for scoping the locators, but that CSS class is ours, and if it changes, we only need to change it in one spot, namely VideoSection.*

IMO, worth making this change in the other tests, too.

File: tests-e2e/util/VideoSection.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import { Page, Locator, FrameLocator, expect } from "@playwright/test"
+
+class VideoSection {
```
Comment (ChristopherChudzicki) [inline tests-e2e/util/VideoSection.ts:3]: **Requested Change:** Could you merge this with `VideoElement`? It got added in #1138 

Be sure to look at the existing methods. E.g., I think `VideoElement.within` would replace the `.videoSection` method here. ("within" is probably a better name)

File: tests-e2e/util/VideoSection.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  constructor(page: Page) {
+    this.page = page
+  }
+
+  /**
+   * Returns a playwright locator for the video section of the page, containing
+   * the video player itself as well as transcript / expandable tabs.
+   *
+   * @fakhar goal here is to scope other the selectors to this part of the DOM.
```
Comment (ChristopherChudzicki) [inline tests-e2e/util/VideoSection.ts:14]: **Requested change:** Could you remove the `@fakhar` comment?

File: tests-e2e/util/index.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 export { default as CoursePage } from "./CoursePage"
+export { default as VideoSection } from "./VideoSection"
 export { default as WwwPage } from "./WwwPage"
```
Comment (ChristopherChudzicki) [inline tests-e2e/util/index.ts:3]: Ahh, the recently added VideoElement did not get added to these exports, or you probably would have noticed it. Oops.

File: tests-e2e/ocw-ci-test-course/shortcodes.spec.ts [hunk 6]
```diff
@@ -37,3 +37,68 @@ test("Related resources link to correct page", async ({ page }) => {
   const href = await resourceLink.getAttribute("href")
   expect(href).toBe("/courses/ocw-ci-test-course/resources/example_pdf/")
 })
+
+test("Start and end times does not exist", async ({ page }) => {
```
Comment (ChristopherChudzicki) [inline tests-e2e/ocw-ci-test-course/shortcodes.spec.ts:41]: **Requested Change:** Could you move these tests to `video.spec.ts`? They aren't really shortocde tests.

**Review by ChristopherChudzicki — CHANGES_REQUESTED** [review]
Looks good, a few small organization requests

**Review by ChristopherChudzicki — CHANGES_REQUESTED** [review]
Looks good, a few small organization requests

**Review by ChristopherChudzicki — APPROVED** [review]
👍 Looks good. One suggestion, but up to you.

**Review by ChristopherChudzicki — APPROVED** [review]
👍 Looks good. One suggestion, but up to you.

**Pull request merged**